x/withdraw/keeper/ibc_callbacks.go

```diff
@@ -0,0 +1,206 @@
+		)
+	}
+
+	// change the bech32 human readable prefix (HRP) of the sender to `evmos1`
+	sender := sdk.AccAddress(senderBz)
+
+	// obtain the evmos recipient address
+	recipient, err := sdk.AccAddressFromBech32(data.Receiver)
+	if err != nil {
+		return channeltypes.NewErrorAcknowledgement(
+			sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid receiver address %s", err.Error()).Error(),
+		)
+	}
+
+	// get the recipient account
+	account := k.accountKeeper.GetAccount(ctx, recipient)
+
+	// NOTE: check if the recipient pubkey is a supported key, if it is,
+	// return the original ACK
+
+	// 1.a. no recipient account or no pubkey -> no way to determine the type of address
+	// ==> return success ACK
+	if account != nil &&
+		account.GetPubKey() != nil &&
+		// 1.b: pubkey is eth_secp256k1 (valid ethereum address) or other supported keys
+		// ==> return success ACK
+		(account.GetPubKey().Type() == ethsecp256k1.KeyType ||
+			account.GetPubKey().Type() == (&multisig.LegacyAminoPubKey{}).Type() ||
+			account.GetPubKey().Type() == (&ed25519.PubKey{}).Type()) {
+		return ack
+	}
+
+	// case 2: sender ≠ recipient and and recipient key type is not supported
+	// Because the destination channel is authorized and not from an EVM chain,
+	// only the secp256k1 keys are supported in the destination
+
+	if !sender.Equals(recipient) {
+		logger.Debug(
+			"rejected IBC transfer to 'secp256k1' key address",
+			"sender", data.Sender,
+			"recipient", data.Receiver,
+			"source-channel", packet.SourceChannel,
+			"destination-channel", packet.DestinationChannel,
+		)
+
+		return channeltypes.NewErrorAcknowledgement(
+			sdkerrors.Wrapf(evmos.ErrKeyTypeNotSupported, "receiver address %s is not a valid ethereum address", data.Receiver).Error(),
+		)
+	}
+
+	// transfer the balance back to the sender address
+	srcPort := packet.SourcePort
+	srcChannel := packet.SourceChannel
+	balances := sdk.Coins{}
+
+	k.bankKeeper.IterateAccountBalances(ctx, recipient, func(coin sdk.Coin) (stop bool) {
+		if coin.IsZero() {
+			// continue
+			return false
+		}
+
+		// we only transfer IBC tokens back to their respective source chains
+		if strings.HasPrefix(coin.Denom, "ibc/") {
+			srcPort, srcChannel, err = k.GetIBCDenomSource(ctx, coin.Denom, data.Sender)
+			if err != nil {
+				logger.Error(
+					"failed to get the IBC full denom path of source chain",
+					"error", err.Error(),
+				)
+				return true // stop iteration
+			}
+
+			// NOTE: only withdraw the IBC tokens from the source channel
+			if packet.SourcePort != srcPort || packet.SourceChannel != srcChannel {
+				// reset to the original values
+				srcPort = packet.SourcePort
+				srcChannel = packet.SourceChannel
+				// continue
+				return false
+			}
+		}
+
+		// Native tokens will be transferred to the authorized source chain to unstuck them
+
+		// NOTE: should we get the timeout from the channel consensus state?
+		timeout := uint64(ctx.BlockTime().Add(time.Hour).UnixNano())
```

Comment:
This is tricky. We have run into this issue as well. Don't use the consensus state because it may become unreliable if updates slow down. Maybe do a longer timeout than 1 hour? The tokens are unaccessible without all this code anyways, so if they need to become timed out (it doesn't matter to the IBC token sender) when this occurs